File: src/python_run/tests/test_zero_shot_e2e.py
"""
E2E zero-shot inference tests using the real test ONNX model.

Tests run direct ONNX inference (bypassing PiperVoice) so that the
phoneme_type='espeak' config in zero-shot-test.onnx.json is not a concern.

Model files are expected at:
  <repo-root>/test/models/zero-shot-test.onnx
  <repo-root>/test/models/zero-shot-test.onnx.json
  <repo-root>/test/models/test_speaker.npy
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest

# ---------------------------------------------------------------------------
# onnxruntime availability guard
# ---------------------------------------------------------------------------
try:
    import onnxruntime  # noqa: F401

    _ORT_AVAILABLE = True
except ImportError:
    _ORT_AVAILABLE = False

ort_skip = pytest.mark.skipif(
    not _ORT_AVAILABLE,
    reason="onnxruntime is not installed",
)

# ---------------------------------------------------------------------------
# Locate model files relative to the repo root
# ---------------------------------------------------------------------------
_REPO_ROOT = Path(__file__).parents[3]  # .../piper-plus-zero-shot/
_MODEL_DIR = _REPO_ROOT / "test" / "models"
_MODEL_PATH = _MODEL_DIR / "zero-shot-test.onnx"
_SPEAKER_NPY = _MODEL_DIR / "test_speaker.npy"

_models_present = _MODEL_PATH.exists() and _SPEAKER_NPY.exists()
model_skip = pytest.mark.skipif(
    not _models_present,
    reason=f"Test model files not found under {_MODEL_DIR}",
)

# Note: These tests use raw onnxruntime.InferenceSession (not PiperVoice),
# so PIPER_DISABLE_WARMUP / PIPER_DISABLE_CACHE have no effect here. They
# are intentionally NOT set at module load to avoid polluting the environment
# for unrelated tests (e.g. test_config_fallback.TestVoiceInlineWarmup).


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

def _load_session() -> "onnxruntime.InferenceSession":
    """Load the zero-shot test ONNX model on CPU."""
    return onnxruntime.InferenceSession(
        str(_MODEL_PATH),
        providers=["CPUExecutionProvider"],
    )


def _base_inputs(phoneme_length: int = 40) -> dict:
    """Return the minimal dict of numpy arrays expected by zero-shot-test.onnx."""
    phoneme_ids = np.zeros((1, phoneme_length), dtype=np.int64)
    phoneme_ids[0, 0] = 1   # BOS
    phoneme_ids[0, -1] = 2  # EOS
    return {
        "input": phoneme_ids,
        "input_lengths": np.array([phoneme_length], dtype=np.int64),
        "scales": np.array([0.4, 1.0, 0.5], dtype=np.float32),
        "prosody_features": np.zeros((1, phoneme_length, 3), dtype=np.int64),
    }


def _run(session, embedding: np.ndarray, phoneme_length: int = 40) -> np.ndarray:
    """Run a single forward pass and return the audio tensor (float32)."""
    inputs = _base_inputs(phoneme_length)
    inputs["speaker_embedding"] = embedding.reshape(1, 192).astype(np.float32)
    return session.run(None, inputs)[0]  # shape [1, 1, T]


# ---------------------------------------------------------------------------
# Test class
# ---------------------------------------------------------------------------

@ort_skip
@model_skip
class TestZeroShotE2E:
    """End-to-end inference tests against zero-shot-test.onnx."""

    @pytest.mark.inference
    def test_e2e_zero_shot_produces_audio(self):
        """Inference with a real speaker embedding produces non-empty, finite audio."""
        session = _load_session()
        embedding = np.load(str(_SPEAKER_NPY))  # shape (192,) float32

        audio = _run(session, embedding)

        # Shape: [batch=1, channels=1, time>0]
        assert audio.ndim == 3, f"Expected 3-D output, got shape {audio.shape}"
        assert audio.shape[0] == 1, "Batch dimension should be 1"
        assert audio.shape[2] > 0, "Audio time dimension must be non-zero"

        # All samples must be finite (no NaN / Inf)
        assert np.all(np.isfinite(audio)), "Audio output contains non-finite values"

        # At least some non-zero samples (not silent)
        assert np.any(audio != 0.0), "Audio output is all zeros"

    @pytest.mark.inference
    def test_e2e_zero_shot_embedding_affects_output(self):
        """Two different speaker embeddings produce different audio outputs."""
        session = _load_session()

        embedding1 = np.load(str(_SPEAKER_NPY)).astype(np.float32)

        # Construct a clearly different embedding (independent random vector)
        rng = np.random.default_rng(seed=12345)
        embedding2 = rng.standard_normal(192).astype(np.float32)

        # Normalise so both embeddings have the same L2 norm
        embedding2 /= np.linalg.norm(embedding2) + 1e-8

        audio1 = _run(session, embedding1)
        audio2 = _run(session, embedding2)

        assert audio1.shape == audio2.shape, (
            f"Shapes differ: {audio1.shape} vs {audio2.shape}"
        )

        # The outputs must differ by at least some margin
        assert not np.allclose(audio1, audio2, atol=1e-6), (
            "Two different speaker embeddings produced identical audio — "
            "speaker conditioning appears to have no effect"
        )

    @pytest.mark.inference
    def test_e2e_zero_shot_zero_embedding(self):
        """A zero-vector speaker embedding does not crash and returns finite audio."""
        session = _load_session()
        zero_embedding = np.zeros(192, dtype=np.float32)

        # Should not raise any exception
        audio = _run(session, zero_embedding)

        assert audio.ndim == 3, f"Expected 3-D output, got shape {audio.shape}"
        assert audio.shape[2] > 0, "Audio time dimension must be non-zero"
        assert np.all(np.isfinite(audio)), (
            "Audio output contains non-finite values when using zero embedding"
        )

File: src/rust/piper-core/tests/test_zero_shot_e2e.rs
//! End-to-end zero-shot TTS inference tests.
//!
//! These tests load the real ONNX model from `test/models/zero-shot-test.onnx`
//! and exercise `OnnxEngine::synthesize()` with a `speaker_embedding` input,
//! verifying that the engine:
//!
//! - produces non-empty, finite audio samples
//! - produces different audio for different speaker embeddings
//! - does not panic on an all-zeros embedding
//!
//! The tests are gated on the `onnx` feature and gracefully skip when the
//! model files are absent (via `assets_available()`).  Run with:
//!
//! ```sh
//! cargo test --test test_zero_shot_e2e --features onnx
//! ```

#![cfg(feature = "onnx")]

use std::path::PathBuf;

use piper_plus::config::VoiceConfig;
use piper_plus::{OnnxEngine, SynthesisRequest};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Resolve a path relative to the repository root.
///
/// `CARGO_MANIFEST_DIR` for `piper-core` is `src/rust/piper-core`.
/// Three `.parent()` calls reach the repo root.
fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap() // piper-core -> src/rust
        .parent()
        .unwrap() // src/rust -> src
        .parent()
        .unwrap() // src -> repo root
        .to_path_buf()
}

fn model_path() -> PathBuf {
    repo_root()
        .join("test")
        .join("models")
        .join("zero-shot-test.onnx")
}

fn config_path() -> PathBuf {
    repo_root()
        .join("test")
        .join("models")
        .join("zero-shot-test.onnx.json")
}

fn npy_path() -> PathBuf {
    repo_root()
        .join("test")
        .join("models")
        .join("test_speaker.npy")
}

/// Return `true` only when all three test asset files exist.
fn assets_available() -> bool {
    model_path().exists() && config_path().exists() && npy_path().exists()
}

/// Load an `OnnxEngine` from the test model.
fn load_engine() -> OnnxEngine {
    let config =
        VoiceConfig::load(&config_path()).expect("Failed to load VoiceConfig from test config");
    OnnxEngine::load(&model_path(), &config, "cpu")
        .expect("Failed to load OnnxEngine from test model")
}

/// Parse a NumPy v1.0 `.npy` file containing a 1-D float32 array.
///
/// Layout: `\x93NUMPY` (6 B) + major (1 B) + minor (1 B) +
///         header_len (2 B LE) + header (header_len B) + data
///
/// Returns the raw `f32` values.
fn load_npy_f32(path: &std::path::Path) -> Vec<f32> {
    let bytes =
        std::fs::read(path).unwrap_or_else(|e| panic!("Failed to read {}: {}", path.display(), e));

    // Verify the NumPy magic bytes.
    assert!(
        bytes.starts_with(b"\x93NUMPY"),
        "File does not start with NumPy magic"
    );

    // Bytes 8-9 (LE u16) encode the length of the ASCII header dict.
    let header_len = u16::from_le_bytes([bytes[8], bytes[9]]) as usize;
    let data_offset = 10 + header_len;

    let data_bytes = &bytes[data_offset..];
    assert_eq!(
        data_bytes.len() % 4,
        0,
        "Data section length is not a multiple of 4"
    );

    data_bytes
        .chunks_exact(4)
        .map(|c| f32::from_le_bytes([c[0], c[1], c[2], c[3]]))
        .collect()
}

/// Build a minimal `SynthesisRequest` long enough to avoid the short-text
/// padding path (MIN_PHONEME_IDS = 40).
///
/// We use IDs [1, 8, 8, …, 8, 2] where 1 = BOS, 2 = EOS, 8 = mid phoneme.
fn make_request(embedding: Option<Vec<f32>>) -> SynthesisRequest {
    const N: usize = 50;
    let mut ids = vec![8i64; N];
    ids[0] = 1; // BOS
    ids[N - 1] = 2; // EOS

    SynthesisRequest {
        phoneme_ids: ids,
        speaker_embedding: embedding,
        noise_scale: 0.4,
        noise_w: 0.5,
        length_scale: 1.0,
        ..SynthesisRequest::default()
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

/// Load the zero-shot model and synthesize audio with a real speaker embedding
/// loaded from `test_speaker.npy`.
///
/// Verifies:
/// - Output audio is non-empty.
/// - Every sample is finite (no NaN / ±Inf).
#[test]
fn test_zero_shot_inference_produces_audio() {
    if !assets_available() {
        eprintln!("SKIP: test model assets not found");
        return;
    }

    let embedding = load_npy_f32(&npy_path());
    assert_eq!(embedding.len(), 192, "Expected 192-dimensional embedding");

    let mut engine = load_engine();
    let request = make_request(Some(embedding));
    let result = engine
        .synthesize(&request)
        .expect("synthesize() should not fail");

    assert!(!result.audio.is_empty(), "Audio output must be non-empty");
    assert!(
        result.audio.iter().all(|&s| (s as f32).is_finite()),
        "All audio samples must be finite"
    );
    assert!(result.sample_rate > 0, "sample_rate must be positive");
    assert!(
        result.audio_seconds > 0.0,
        "audio_seconds must be positive, got {}",
        result.audio_seconds
    );
}

/// Verify that two different speaker embeddings produce different audio output.
///
/// A zero-shot model must condition its output on the speaker embedding;
/// if both outputs were identical the speaker encoder would be non-functional.
#[test]
fn test_zero_shot_different_embeddings_produce_different_audio() {
    if !assets_available() {
        eprintln!("SKIP: test model assets not found");
        return;
    }

    let emb1 = load_npy_f32(&npy_path());
    assert_eq!(emb1.len(), 192);

    // Create a clearly different embedding by negating all values.
    let emb2: Vec<f32> = emb1.iter().map(|v| -v).collect();

    let mut engine = load_engine();

    let result1 = engine
        .synthesize(&make_request(Some(emb1)))
        .expect("synthesize() with emb1 should not fail");

    let result2 = engine
        .synthesize(&make_request(Some(emb2)))
        .expect("synthesize() with emb2 should not fail");

    // The outputs must differ in at least one sample.  We compare a prefix
    // whose length is the minimum of both outputs to handle variable-length audio.
    let min_len = result1.audio.len().min(result2.audio.len());
    assert!(min_len > 0, "Both outputs must be non-empty");

    let differ = result1.audio[..min_len]
        .iter()
        .zip(result2.audio[..min_len].iter())
        .any(|(a, b)| a != b);

    assert!(
        differ,
        "Audio from different speaker embeddings must differ"
    );
}

/// Verify that synthesizing with an all-zeros speaker embedding does not panic.
///
/// The model logs a warning about the missing embedding and falls back to zeros
/// internally, so passing zeros explicitly is a valid (if degenerate) use-case.
#[test]
fn test_zero_shot_zero_embedding_does_not_crash() {
    if !assets_available() {
        eprintln!("SKIP: test model assets not found");
        return;
    }

    let zero_emb = vec![0.0f32; 192];

    let mut engine = load_engine();
    let result = engine.synthesize(&make_request(Some(zero_emb)));

    // The call must not panic.  A successful result is ideal; an `Err` is
    // also acceptable (some models reject degenerate inputs), but a panic
    // (including an unwrap inside OnnxEngine) is a failure.
    match result {
        Ok(r) => {
            // If synthesis succeeded, basic sanity checks apply.
            assert!(
                r.audio.is_empty() || r.audio.iter().all(|&s| (s as f32).is_finite()),
                "Audio samples must be finite when zero-embedding synthesis succeeds"
            );
        }
        Err(e) => {
            // A model-level error is acceptable; we only forbid panics.
            eprintln!("zero-embedding synthesis returned Err (acceptable): {e}");
        }
    }
}
